File: WinForms/DevExpress.OutlookInspiredApp.Win/VB/DevExpress.OutlookInspiredApp.Win/Modules/Tasks/Tasks.vb
Imports System
Imports System.Windows.Forms
Imports DevExpress.DevAV.Presenters
Imports DevExpress.DevAV.ViewModels
Imports DevExpress.XtraBars
Imports DevExpress.XtraGantt
Imports DevExpress.XtraLayout.Utils

Namespace DevExpress.DevAV.Modules

    Public Partial Class Tasks
        Inherits BaseModuleControl
        Implements IRibbonModule

        Private _CollectionUIViewModel As CollectionUIViewModel, _CollectionPresenter As TasksCollectionPresenterType

        Public Sub New()
            MyBase.New(GetType(TaskCollectionViewModel))
            InitializeComponent()
            ganttControl.EnableDynamicLoading = False
            ganttControl.TreeListMappings.KeyFieldName = "Id"
            ganttControl.TreeListMappings.ParentFieldName = "ParentId"
            ganttControl.ChartMappings.StartDateFieldName = "StartDate"
            ganttControl.ChartMappings.FinishDateFieldName = "DueDate"
            ganttControl.ChartMappings.ProgressFieldName = "Completion"
            ganttControl.ChartMappings.PredecessorsFieldName = "Predecessors"
            CollectionUIViewModel = Mvvm.POCO.ViewModelSource.Create(Of CollectionUIViewModel)()
            CollectionPresenter = CreateCollectionPresenter()
            CollectionPresenter.ReloadEntities(mvvmContext)
            BindCommands()
            InitViewKind()
            InitViewLayout()
        End Sub

        Protected Overrides Sub OnDisposing()
            CollectionPresenter.Dispose()
            MyBase.OnDisposing()
        End Sub

#Region "ViewKind"
        Private Sub InitViewKind()
            biResetView.BindCommand(Sub() CollectionUIViewModel.ResetView(), CollectionUIViewModel)
        End Sub

#End Region
        Private Sub InitViewLayout()
            AddHandler CollectionUIViewModel.ViewLayoutChanged, AddressOf Me.ViewModel_ViewLayoutChanged
            bmiHorizontalLayout.BindCommand(Sub() CollectionUIViewModel.ShowHorizontalLayout(), CollectionUIViewModel)
            bmiVerticalLayout.BindCommand(Sub() CollectionUIViewModel.ShowVerticalLayout(), CollectionUIViewModel)
            bmiHideDetail.BindCommand(Sub() CollectionUIViewModel.HideDetail(), CollectionUIViewModel)
            InitViewMenuItems()
        End Sub

        Private Sub ViewModel_ViewLayoutChanged(ByVal sender As Object, ByVal e As EventArgs)
            Dim detailHidden As Boolean = CollectionUIViewModel.IsDetailHidden
            roundedSkinPanelPnlView.Visible = Not detailHidden
            If Not detailHidden Then
                roundedSkinPanelPnlView.Dock = If(CollectionUIViewModel.IsHorizontalLayout, DockStyle.Right, DockStyle.Bottom)
                If CollectionUIViewModel.IsHorizontalLayout Then
                    roundedSkinPanelPnlView.Width = ScaleDPI.ScaleHorizontal(400)
                Else
                    roundedSkinPanelPnlView.Height = ScaleDPI.ScaleHorizontal(400)
                End If
            End If
        End Sub

        Public ReadOnly Property ViewModel As TaskCollectionViewModel
            Get
                Return GetViewModel(Of TaskCollectionViewModel)()
            End Get
        End Property

        Protected Property CollectionUIViewModel As CollectionUIViewModel
            Get
                Return _CollectionUIViewModel
            End Get

            Private Set(ByVal value As CollectionUIViewModel)
                _CollectionUIViewModel = value
            End Set
        End Property

        Protected Property CollectionPresenter As TasksCollectionPresenterType
            Get
                Return _CollectionPresenter
            End Get

            Private Set(ByVal value As TasksCollectionPresenterType)
                _CollectionPresenter = value
            End Set
        End Property

        Private ReadOnly Property OptionsSplitter As GanttControlOptionsSplitter
            Get
                If ganttControl IsNot Nothing Then Return ganttControl.OptionsSplitter
                Return Nothing
            End Get
        End Property

        Protected Overridable Function CreateCollectionPresenter() As TasksCollectionPresenterType
            Return New TasksCollectionPresenterType(ganttControl, ViewModel, AddressOf UpdateEntitiesCountRelatedUI)
        End Function

        Protected Overrides Sub OnInitServices()
            mvvmContext.RegisterService("View Settings", New ViewSettingsDialogDocumentManagerService(Function() CollectionUIViewModel))
            mvvmContext.RegisterService(New DetailFormDocumentManagerService(ModuleType.TaskEditView))
        End Sub

        Private Sub UpdateEntitiesCountRelatedUI(ByVal count As Integer)
            hiItemsCount.Caption = String.Format("Records: {0}", count)
        End Sub

        Private ReadOnly Property Ribbon As XtraBars.Ribbon.RibbonControl Implements IRibbonModule.Ribbon
            Get
                Return ribbonControl
            End Get
        End Property

        Private taskView As TaskView

        Protected Overrides Sub OnLoad(ByVal e As EventArgs)
            MyBase.OnLoad(e)
            Dim moduleLocator = GetService(Of Services.IModuleLocator)()
            taskView = TryCast(moduleLocator.GetModule(ModuleType.TaskView), TaskView)
            EnsureModuleViewModel(taskView, ViewModel, ViewModel.SelectedEntityKey)
            taskView.Dock = DockStyle.Fill
            taskView.Parent = roundedSkinPanelPnlView
        End Sub

        Private Sub OnAllowResizeDownChanged(ByVal sender As Object, ByVal e As ItemClickEventArgs)
            Dim checkItem = TryCast(e.Item, BarButtonItem)
            If OptionsSplitter IsNot Nothing AndAlso checkItem IsNot Nothing Then OptionsSplitter.AllowResize = checkItem.Down
        End Sub

        Private Sub bciPanelVisibilityDefault_PanelVisibilityItemClick(ByVal sender As Object, ByVal e As ItemClickEventArgs)
            ApplyPanelVisibility(e.Item.ToString())
        End Sub

        Private Sub bciPanelVisibilityChart_PanelVisibilityItemClick(ByVal sender As Object, ByVal e As ItemClickEventArgs)
            ApplyPanelVisibility(e.Item.ToString())
        End Sub

        Private Sub bciPanelVisibilityTree_PanelVisibilityItemClick(ByVal sender As Object, ByVal e As ItemClickEventArgs)
            ApplyPanelVisibility(e.Item.ToString())
        End Sub

        Private Sub bciPanelVisibilityBoth_PanelVisibilityItemClick(ByVal sender As Object, ByVal e As ItemClickEventArgs)
            ApplyPanelVisibility(e.Item.ToString())
        End Sub

        Private Sub bciFixedPanelDefault_FixedPanelItemClick(ByVal sender As Object, ByVal e As ItemClickEventArgs)
            ApplyFixedPanel(e.Item.ToString())
        End Sub

        Private Sub bciFixedPanelChart_FixedPanelItemClick(ByVal sender As Object, ByVal e As ItemClickEventArgs)
            ApplyFixedPanel(e.Item.ToString())
        End Sub

        Private Sub bciFixedPanelNone_FixedPanelItemClick(ByVal sender As Object, ByVal e As ItemClickEventArgs)
            ApplyFixedPanel(e.Item.ToString())
        End Sub

        Private Sub bciFixedPanelTree_FixedPanelItemClick(ByVal sender As Object, ByVal e As ItemClickEventArgs)
            ApplyFixedPanel(e.Item.ToString())
        End Sub

        Private Sub ApplyPanelVisibility(ByVal name As String)
            If OptionsSplitter IsNot Nothing Then OptionsSplitter.PanelVisibility = CType([Enum].Parse(GetType(GanttPanelVisibility), name), GanttPanelVisibility)
        End Sub

        Private Sub ApplyFixedPanel(ByVal name As String)
            If OptionsSplitter IsNot Nothing Then OptionsSplitter.FixedPanel = CType([Enum].Parse(GetType(GanttFixedPanel), name), GanttFixedPanel)
        End Sub

        Private Sub InitViewMenuItems()
            biAllowResize.Down = OptionsSplitter.AllowResize
            Select Case OptionsSplitter.FixedPanel
                Case GanttFixedPanel.Default
                    bciFixedPanelDefault.Checked = True
                Case GanttFixedPanel.Tree
                    bciFixedPanelTree.Checked = True
                Case GanttFixedPanel.Chart
                    bciFixedPanelChart.Checked = True
                Case GanttFixedPanel.None
                    bciFixedPanelNone.Checked = True
            End Select

            Select Case OptionsSplitter.PanelVisibility
                Case GanttPanelVisibility.Default
                    bciPanelVisibilityDefault.Checked = True
                Case GanttPanelVisibility.Tree
                    bciPanelVisibilityTree.Checked = True
                Case GanttPanelVisibility.Chart
                    bciPanelVisibilityChart.Checked = True
                Case GanttPanelVisibility.Both
                    bciPanelVisibilityBoth.Checked = True
            End Select
        End Sub

        Private Sub BindCommands()
            biNewTask.BindCommand(Sub() ViewModel.[New](), ViewModel)
            biEdit.BindCommand(Sub(e) ViewModel.Edit(e), ViewModel, Function() ViewModel.SelectedEntity)
            biDelete.BindCommand(Sub(e) ViewModel.Delete(e), ViewModel, Function() ViewModel.SelectedEntity)
            biViewSettings.BindCommand(Sub() ViewModel.ShowViewSettings(), ViewModel)
        End Sub
    End Class
End Namespace

File: WinForms/DevExpress.OutlookInspiredApp.Win/CS/DevExpress.OutlookInspiredApp.Win/Modules/Tasks/Tasks.cs
using System;
using System.Windows.Forms;
using DevExpress.DevAV.Presenters;
using DevExpress.DevAV.ViewModels;
using DevExpress.XtraBars;
using DevExpress.XtraGantt;
using DevExpress.XtraLayout.Utils;

namespace DevExpress.DevAV.Modules {
    public partial class Tasks : BaseModuleControl, IRibbonModule {
        public Tasks() : base(typeof(TaskCollectionViewModel)) {
            InitializeComponent();
            ganttControl.EnableDynamicLoading = false;
            ganttControl.TreeListMappings.KeyFieldName = "Id";
            ganttControl.TreeListMappings.ParentFieldName = "ParentId";
            ganttControl.ChartMappings.StartDateFieldName = "StartDate";
            ganttControl.ChartMappings.FinishDateFieldName = "DueDate";
            ganttControl.ChartMappings.ProgressFieldName = "Completion";
            ganttControl.ChartMappings.PredecessorsFieldName = "Predecessors";
            

            CollectionUIViewModel = Mvvm.POCO.ViewModelSource.Create<CollectionUIViewModel>();
            CollectionPresenter = CreateCollectionPresenter();
            CollectionPresenter.ReloadEntities(mvvmContext);
            BindCommands();

            InitViewKind();
            InitViewLayout();
        }
        protected override void OnDisposing() {
            CollectionPresenter.Dispose();
            base.OnDisposing();
        }
        #region ViewKind
        void InitViewKind() {
            biResetView.BindCommand(() => CollectionUIViewModel.ResetView(), CollectionUIViewModel);
        }
        #endregion

        void InitViewLayout() {
            CollectionUIViewModel.ViewLayoutChanged += ViewModel_ViewLayoutChanged;
            bmiHorizontalLayout.BindCommand(() => CollectionUIViewModel.ShowHorizontalLayout(), CollectionUIViewModel);
            bmiVerticalLayout.BindCommand(() => CollectionUIViewModel.ShowVerticalLayout(), CollectionUIViewModel);
            bmiHideDetail.BindCommand(() => CollectionUIViewModel.HideDetail(), CollectionUIViewModel);
            InitViewMenuItems();
        }
        void ViewModel_ViewLayoutChanged(object sender, EventArgs e) {
            bool detailHidden = CollectionUIViewModel.IsDetailHidden;
            roundedSkinPanelPnlView.Visible = !detailHidden;
            if(!detailHidden) {
                roundedSkinPanelPnlView.Dock = CollectionUIViewModel.IsHorizontalLayout ? DockStyle.Right : DockStyle.Bottom;
                if(CollectionUIViewModel.IsHorizontalLayout) {
                    roundedSkinPanelPnlView.Width = ScaleDPI.ScaleHorizontal(400);
                }
                else {
                    roundedSkinPanelPnlView.Height = ScaleDPI.ScaleHorizontal(400);
                }
            }
        }
        public TaskCollectionViewModel ViewModel {
            get { return GetViewModel<TaskCollectionViewModel>(); }
        }
        protected CollectionUIViewModel CollectionUIViewModel { get; private set; }
        protected TasksCollectionPresenter CollectionPresenter { get; private set; }
        GanttControlOptionsSplitter OptionsSplitter {
            get {
                if(ganttControl != null)
                    return ganttControl.OptionsSplitter;
                return null;
            }
        }
        protected virtual TasksCollectionPresenter CreateCollectionPresenter() {
            return new TasksCollectionPresenter(ganttControl, ViewModel, UpdateEntitiesCountRelatedUI);
        }
        protected override void OnInitServices() {
            mvvmContext.RegisterService("View Settings", new ViewSettingsDialogDocumentManagerService(() => CollectionUIViewModel));
            mvvmContext.RegisterService(new DetailFormDocumentManagerService(ModuleType.TaskEditView));
        }
        void UpdateEntitiesCountRelatedUI(int count) {
            hiItemsCount.Caption = string.Format("Records: {0}", count);
        }
        XtraBars.Ribbon.RibbonControl IRibbonModule.Ribbon {
            get { return ribbonControl; }
        }
        TaskView taskView;
        protected override void OnLoad(EventArgs e) {
            base.OnLoad(e);
            var moduleLocator = GetService<Services.IModuleLocator>();
            taskView = moduleLocator.GetModule(ModuleType.TaskView) as TaskView;
            ViewModelHelper.EnsureModuleViewModel(taskView, ViewModel, ViewModel.SelectedEntityKey);
            taskView.Dock = DockStyle.Fill;
            taskView.Parent = roundedSkinPanelPnlView;
        }
        void OnAllowResizeDownChanged(object sender, ItemClickEventArgs e) {
            var checkItem = e.Item as BarButtonItem;
            if(OptionsSplitter != null && checkItem != null)
                OptionsSplitter.AllowResize = checkItem.Down;
        }
        void bciPanelVisibilityDefault_PanelVisibilityItemClick(object sender, ItemClickEventArgs e) {
            ApplyPanelVisibility(e.Item.ToString());
        }
        void bciPanelVisibilityChart_PanelVisibilityItemClick(object sender, ItemClickEventArgs e) {
            ApplyPanelVisibility(e.Item.ToString());
        }
        void bciPanelVisibilityTree_PanelVisibilityItemClick(object sender, ItemClickEventArgs e) {
            ApplyPanelVisibility(e.Item.ToString());
        }
        void bciPanelVisibilityBoth_PanelVisibilityItemClick(object sender, ItemClickEventArgs e) {
            ApplyPanelVisibility(e.Item.ToString());
        }
        void bciFixedPanelDefault_FixedPanelItemClick(object sender, ItemClickEventArgs e) {
            ApplyFixedPanel(e.Item.ToString());
        }
        void bciFixedPanelChart_FixedPanelItemClick(object sender, ItemClickEventArgs e) {
            ApplyFixedPanel(e.Item.ToString());
        }
        void bciFixedPanelNone_FixedPanelItemClick(object sender, ItemClickEventArgs e) {
            ApplyFixedPanel(e.Item.ToString());
        }
        void bciFixedPanelTree_FixedPanelItemClick(object sender, ItemClickEventArgs e) {
            ApplyFixedPanel(e.Item.ToString());
        }
        void ApplyPanelVisibility(string name) {
            if(OptionsSplitter != null)
                OptionsSplitter.PanelVisibility = (GanttPanelVisibility)Enum.Parse(typeof(GanttPanelVisibility), name);
        }
        void ApplyFixedPanel(string name) {
            if(OptionsSplitter != null)
                OptionsSplitter.FixedPanel = (GanttFixedPanel)Enum.Parse(typeof(GanttFixedPanel), name);
        }
        void InitViewMenuItems() {
            biAllowResize.Down = OptionsSplitter.AllowResize;
            switch(OptionsSplitter.FixedPanel) {
                case GanttFixedPanel.Default:
                    bciFixedPanelDefault.Checked = true;
                    break;
                case GanttFixedPanel.Tree:
                    bciFixedPanelTree.Checked = true;
                    break;
                case GanttFixedPanel.Chart:
                    bciFixedPanelChart.Checked = true;
                    break;
                case GanttFixedPanel.None:
                    bciFixedPanelNone.Checked = true;
                    break;
            }
            switch(OptionsSplitter.PanelVisibility) {
                case GanttPanelVisibility.Default:
                    bciPanelVisibilityDefault.Checked = true;
                    break;
                case GanttPanelVisibility.Tree:
                    bciPanelVisibilityTree.Checked = true;
                    break;
                case GanttPanelVisibility.Chart:
                    bciPanelVisibilityChart.Checked = true;
                    break;
                case GanttPanelVisibility.Both:
                    bciPanelVisibilityBoth.Checked = true;
                    break;
            }
        }
        void BindCommands() {
            biNewTask.BindCommand(() => ViewModel.New(), ViewModel);
            biEdit.BindCommand((e) => ViewModel.Edit(e), ViewModel, () => ViewModel.SelectedEntity);
            biDelete.BindCommand((e) => ViewModel.Delete(e), ViewModel, () => ViewModel.SelectedEntity);
            biViewSettings.BindCommand(() => ViewModel.ShowViewSettings(), ViewModel);
        }
    }
}
